client/src/main/java/io/pravega/client/segment/impl/EventSegmentReaderImpl.java

```diff
@@ -54,30 +59,44 @@ public long getOffset() {
      */
     @Override
     @Synchronized
-    public ByteBuffer read(long timeout) throws EndOfSegmentException, SegmentTruncatedException {
+    public ByteBuffer read(long firstByteTimeoutMillis) throws EndOfSegmentException, SegmentTruncatedException {
         long originalOffset = in.getOffset();
-        long traceId = LoggerHelpers.traceEnter(log, "read", in.getSegmentId(), originalOffset, timeout);
+        long traceId = LoggerHelpers.traceEnter(log, "read", in.getSegmentId(), originalOffset, firstByteTimeoutMillis);
         boolean success = false;
+        boolean timeout = false;
         try {
-            ByteBuffer result = readEvent(timeout);
+            ByteBuffer result = readEvent(firstByteTimeoutMillis);
             success = true;
             return result;
+        } catch (TimeoutException e) {
+            timeout = true;
+            log.warn("Timeout observed while trying to read data from Segment store, the read request will be retransmitted. Details: {}", e.getMessage());
+            return null;
         } finally {
-            LoggerHelpers.traceLeave(log, "read", traceId, in.getSegmentId(), originalOffset, timeout, success);
+            LoggerHelpers.traceLeave(log, "read", traceId, in.getSegmentId(), originalOffset, firstByteTimeoutMillis, success);
             if (!success) {
-                in.setOffset(originalOffset);
+                if (timeout) {
+                    in.setOffset(originalOffset, true);
+                } else {
+                    in.setOffset(originalOffset);
+                }
             }
         }
     }
         
-    public ByteBuffer readEvent(long timeout) throws EndOfSegmentException, SegmentTruncatedException {
+    public ByteBuffer readEvent(long firstByteTimeoutMillis) throws EndOfSegmentException, SegmentTruncatedException, TimeoutException {
         headerReadingBuffer.clear();
-        int read = in.read(headerReadingBuffer, timeout);
+        int read = in.read(headerReadingBuffer, firstByteTimeoutMillis);
         if (read == 0) {
+            // a resend will not be triggered in-case of a firstByteTimeout.
             return null;
         }
         while (headerReadingBuffer.hasRemaining()) {
-            in.read(headerReadingBuffer, Long.MAX_VALUE);
+            if (in.read(headerReadingBuffer, READ_TIMEOUT_MS) == 0) {
+                log.warn("Timeout out while trying to read WireCommand header during read of segment {} at offset {}",
```

Comment:
Given three instances of this if and throw, it probably makes sense to have a private method.